compiler/rustc_ast/src/expand/autodiff_attrs.rs

```diff
@@ -0,0 +1,270 @@
+use std::fmt::{self, Display, Formatter};
+use std::str::FromStr;
+
+use crate::expand::typetree::TypeTree;
+use crate::expand::{Decodable, Encodable, HashStable_Generic};
+use crate::ptr::P;
+use crate::{Ty, TyKind};
+
+#[allow(dead_code)]
+#[derive(Clone, Copy, Eq, PartialEq, Encodable, Decodable, Debug, HashStable_Generic)]
+pub enum DiffMode {
+    Inactive,
+    Source,
+    Forward,
+    Reverse,
+    ForwardFirst,
+    ReverseFirst,
+}
+
+pub fn is_rev(mode: DiffMode) -> bool {
+    match mode {
+        DiffMode::Reverse | DiffMode::ReverseFirst => true,
+        _ => false,
+    }
+}
+pub fn is_fwd(mode: DiffMode) -> bool {
+    match mode {
+        DiffMode::Forward | DiffMode::ForwardFirst => true,
+        _ => false,
+    }
+}
+
+impl Display for DiffMode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            DiffMode::Inactive => write!(f, "Inactive"),
+            DiffMode::Source => write!(f, "Source"),
+            DiffMode::Forward => write!(f, "Forward"),
+            DiffMode::Reverse => write!(f, "Reverse"),
+            DiffMode::ForwardFirst => write!(f, "ForwardFirst"),
+            DiffMode::ReverseFirst => write!(f, "ReverseFirst"),
+        }
+    }
+}
+
+pub fn valid_ret_activity(mode: DiffMode, activity: DiffActivity) -> bool {
+    if activity == DiffActivity::None {
+        // Only valid if primal returns (), but we can't check that here.
+        return true;
+    }
+    match mode {
+        DiffMode::Inactive => false,
+        DiffMode::Source => false,
+        DiffMode::Forward | DiffMode::ForwardFirst => {
+            activity == DiffActivity::Dual
+                || activity == DiffActivity::DualOnly
+                || activity == DiffActivity::Const
+        }
+        DiffMode::Reverse | DiffMode::ReverseFirst => {
+            activity == DiffActivity::Const
+                || activity == DiffActivity::Active
+                || activity == DiffActivity::ActiveOnly
+        }
+    }
+}
+fn is_ptr_or_ref(ty: &Ty) -> bool {
+    match ty.kind {
+        TyKind::Ptr(_) | TyKind::Ref(_, _) => true,
+        _ => false,
+    }
+}
+// TODO We should make this more robust to also
+// accept aliases of f32 and f64
+//fn is_float(ty: &Ty) -> bool {
+//    false
+//}
+pub fn valid_ty_for_activity(ty: &P<Ty>, activity: DiffActivity) -> bool {
```

Comment:
This operating on the AST is somewhat sketchy as well (there are *many* alternative ways to name the types -- type aliases, associatedd types, etc), though if this needs to operate as a macro then whatever.

Reply:
Yeah, I think the roadblock/problem here is that the macro seems to need to operate at a AST level, before full type info is available.

Reply:
Exactly. I currently try to error as early as possible, but that means I can't catch all issues.
In case of doubt I allow a combination, knowing that it will probably cause an Enzyme/LLVM panic if illegal.
Over time I'll add more checks in the middle-end where types are available. Maybe at some point
we even want to get rid completely of frontend checks, but that's what we have for now. 

I really want autodiff to be a macro that generates a new function instead of letting users call undefined functions (even though I could support that), to help users/IDEs/etc.
That however means I need to generate the new function before type resolution.

Reply:
> Over time I'll add more checks in the middle-end where types are available. Maybe at some point
we even want to get rid completely of frontend checks, but that's what we have for now.

Yes that makes sense to me, to move the type-specific validations to after type info is available.

In the meantime, can you document this limitation to the validation checks we have in the frontend?

Reply:
yes, added.